Core/Transaction_Handler.php

```diff
@@ -0,0 +1,215 @@
+
+use \Wirecard\Oxid\Model\Transaction;
+
+/**
+ * Class Transaction_Handler
+ * @package Wirecard\Oxid\Core
+ */
+class Transaction_Handler
+{
+    const TRANSACTION_STATUS_SUCCESS = 'success';
+    const TRANSACTION_STATUS_ERROR = 'error';
+
+    /**
+     * @var \Psr\Log\LoggerInterface
+     */
+    private $_oLogger;
+
+    /**
+     * Transaction_Handler constructor.
+     */
+    public function __construct()
+    {
+        $this->_oLogger = Registry::getLogger();
+    }
+
+    /**
+     * @param Transaction $oParentTransaction
+     * @param string      $sActionTitle
+     * @param float       $fAmount
+     *
+     * @return array either [status => success] or [status => error, message => errorMessage]
+     */
+    public function processAction(Transaction $oParentTransaction, String $sActionTitle, float $fAmount = null): array
+    {
+        $oPayment = $this->_getPaymentMethod($oParentTransaction);
+        $oConfig = $oPayment->getConfig();
+        $oOrder = $oParentTransaction->getTransactionOrder();
+
+        $oTransaction = $this->_getTransactionObject($oPayment, $sActionTitle);
+
+        if (empty($oTransaction)) {
+            $this->_oLogger->error("action not implemented", ['no implementation for ' . $sActionTitle]);
+            return self::_error('Action "' . $sActionTitle . '" is not implemented');
+        }
+
+        $sParentTransactionId = $oParentTransaction->wdoxidee_ordertransactions__transactionid->value;
+        $oTransaction->setParentTransactionId($sParentTransactionId);
+
+        if (!is_null($fAmount)) {
+            $sCurrencyName = $oOrder->oxorder__oxcurrency->value;
+            $oCurrency = Registry::getConfig()->getCurrencyObject($sCurrencyName);
+
+            $oTransaction->setAmount(new Amount(
+                Registry::getUtils()->fRound($fAmount, $oCurrency),
+                $sCurrencyName
+            ));
+        }
+
+        // use the backend service to process the action
+        $oBackendService = new BackendService($oConfig, $this->_oLogger);
+        $oResponse = $oBackendService->process($oTransaction, $sActionTitle);
+        $this->_oLogger->debug('response from ' . $sActionTitle . ' action: ' . $oResponse->getRawData());
+
+        return $this->_handleProcessActionResponse($oResponse, $oOrder, $oBackendService);
+    }
+
+    /**
+     * @param Response       $oResponse
+     * @param Order          $oOrder
+     * @param BackendService $oBackendService
+     *
+     * @return array either [status => success] or [status => error, message => errorMessage]
+     */
+    private function _handleProcessActionResponse($oResponse, $oOrder, $oBackendService)
+    {
+        if ($oResponse instanceof FailureResponse) {
+            $sErrors = '';
+
+            foreach ($oResponse->getStatusCollection()->getIterator() as $oItem) {
+                $sErrors .= $oItem->getDescription() . "<br>";
+                $this->_oLogger->error($oItem->getDescription());
```

Comment:
Please, log also `status->getCode()`. We should log both description and status. See OrderController.php

Reply:
Thanks 👍